wled00/FX_2Dfcn.cpp

```diff
@@ -192,29 +197,113 @@ void IRAM_ATTR Segment::setPixelColorXY(int x, int y, uint32_t col)
   y *= groupLength(); // expand to physical pixels
   if (x >= width() || y >= height()) return;  // if pixel would fall out of segment just exit
 
-  uint32_t tmpCol = col;
   for (int j = 0; j < grouping; j++) {   // groupping vertically
     for (int g = 0; g < grouping; g++) { // groupping horizontally
       uint16_t xX = (x+g), yY = (y+j);
       if (xX >= width() || yY >= height()) continue; // we have reached one dimension's end
 
-#ifndef WLED_DISABLE_MODE_BLEND
-      // if blending modes, blend with underlying pixel
-      if (_modeBlend) tmpCol = color_blend(strip.getPixelColorXY(start + xX, startY + yY), col, 0xFFFFU - progress(), true);
-#endif
-
-      strip.setPixelColorXY(start + xX, startY + yY, tmpCol);
+      strip.setPixelColorXY(start + xX, startY + yY, col);
 
       if (mirror) { //set the corresponding horizontally mirrored pixel
-        if (transpose) strip.setPixelColorXY(start + xX, startY + height() - yY - 1, tmpCol);
-        else           strip.setPixelColorXY(start + width() - xX - 1, startY + yY, tmpCol);
+        if (transpose) strip.setPixelColorXY(start + xX, startY + height() - yY - 1, col);
+        else           strip.setPixelColorXY(start + width() - xX - 1, startY + yY, col);
       }
       if (mirror_y) { //set the corresponding vertically mirrored pixel
-        if (transpose) strip.setPixelColorXY(start + width() - xX - 1, startY + yY, tmpCol);
-        else           strip.setPixelColorXY(start + xX, startY + height() - yY - 1, tmpCol);
+        if (transpose) strip.setPixelColorXY(start + width() - xX - 1, startY + yY, col);
+        else           strip.setPixelColorXY(start + xX, startY + height() - yY - 1, col);
       }
       if (mirror_y && mirror) { //set the corresponding vertically AND horizontally mirrored pixel
-        strip.setPixelColorXY(width() - xX - 1, height() - yY - 1, tmpCol);
+        strip.setPixelColorXY(width() - xX - 1, height() - yY - 1, col);
+      }
+    }
+  }
+}
+
+void Segment::render2DTransition() {
+  uint16_t width = virtualWidth();
+  uint16_t height = virtualHeight();
+
+  // Sometimes the timing works out such that "pos" calculated below doesn't hit
+  // all the possible coordinates. This can happen e.g. when the frame rate is low.
+  // As a result, there can be some pixels left in a transitional state. This is a
+  // problem if the target effect doesn't update all the pixels. In order to ensure
+  // there are no artifacts left from the transition, we need to clear the segment
+  // before we render the transition.
+  fill(BLACK);
```

Comment:
Can you think of a better way to handle this?
`fill()` uses `setPixelColor()/setPixelColorXY()`.

Reply:
Curious, what's the problem with that? The goal is to clear all pixels, taking segment grouping, spacing and mirroring into account. Is there a simpler or more efficient way?

Reply:
Well, IMO `fill()` is an effect function and may (in the future) employ different principles of filling a segment. If you look at the definition you'll see that it does not *fill* every pixel in every instance. It may skip pixels if spacing is used.
I think it would be wiser to employ classic buffer cleaning using a `for` loop or `memset()`.

Reply:
~~And then consider user changing spacing on segment without changing anything else. You'll get artifacts remaining.~~

I have checked the code and I actually added segment clearing prior to changing spacing, so all is good.
I had to mentally run through the implementation and it is indeed ok. At first I thought that overlapping segments would be overwritten but now I see that `setPixelColor()` is reused to do the actual painting in `renderTransition()`.